lib/std/fs.zig

```diff
@@ -761,65 +799,80 @@ pub const Dir = struct {
-            creation,
-            w.FILE_NON_DIRECTORY_FILE | w.FILE_SYNCHRONOUS_IO_NONALERT,
-            null,
-            0,
-        );
-        switch (rc) {
-            .SUCCESS => return result,
-            .OBJECT_NAME_INVALID => unreachable,
-            .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
-            .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
-            .NO_MEDIA_IN_DEVICE => return error.NoDevice,
-            .INVALID_PARAMETER => unreachable,
-            .SHARING_VIOLATION => return error.SharingViolation,
-            .ACCESS_DENIED => return error.AccessDenied,
-            .PIPE_BUSY => return error.PipeBusy,
-            .OBJECT_PATH_SYNTAX_BAD => unreachable,
-            .OBJECT_NAME_COLLISION => return error.PathAlreadyExists,
-            else => return w.unexpectedStatus(rc),
+            const path_len_bytes = math.cast(u16, mem.toSliceConst(u16, sub_path_w).len * 2) catch |err| switch (err) {
+                error.Overflow => return error.NameTooLong,
+            };
+            var nt_name = w.UNICODE_STRING{
+                .Length = path_len_bytes,
+                .MaximumLength = path_len_bytes,
+                .Buffer = @intToPtr([*]u16, @ptrToInt(sub_path_w)),
+            };
+            var attr = w.OBJECT_ATTRIBUTES{
+                .Length = @sizeOf(w.OBJECT_ATTRIBUTES),
+                .RootDirectory = if (path.isAbsoluteWindowsW(sub_path_w)) null else self.fd,
+                .Attributes = 0, // Note we do not use OBJ_CASE_INSENSITIVE here.
+                .ObjectName = &nt_name,
+                .SecurityDescriptor = null,
+                .SecurityQualityOfService = null,
+            };
+            var io: w.IO_STATUS_BLOCK = undefined;
+            const share_access = share_access_opt orelse w.FILE_SHARE_WRITE | w.FILE_SHARE_READ | w.FILE_SHARE_DELETE;
+            const rc = w.ntdll.NtCreateFile(
+                &result.handle,
+                access_mask,
+                &attr,
+                &io,
+                null,
+                w.FILE_ATTRIBUTE_NORMAL,
+                share_access,
+                creation,
+                w.FILE_NON_DIRECTORY_FILE | w.FILE_SYNCHRONOUS_IO_NONALERT,
+                null,
+                0,
+            );
+            switch (rc) {
+                .SUCCESS => return result,
+                .OBJECT_NAME_INVALID => unreachable,
+                .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
+                .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
+                .NO_MEDIA_IN_DEVICE => return error.NoDevice,
+                .INVALID_PARAMETER => unreachable,
+                .SHARING_VIOLATION => {
+                    // TODO: check if async or blocking
+                    //return error.SharingViolation
+                    // Sleep so we don't consume a ton of CPU waiting to get lock on file
+                    std.time.sleep(delay);
```

Comment:
Can this be and endless loop?

Reply:
Yes. I decided to go this route because on Linux, the call to `fcntl` with SETLKW will also block until the file can be locked. I figure that we can make a version that returns immediately once async is integrated.